dashboard/lib/sqs/queue_processor.rb

```diff
@@ -0,0 +1,129 @@
+# particular requirements for high scale processing:
+# - The existing ActiveJob worker interface does not support batching.
+# - Shoryuken's custom API supports batching but not throttling, and the
+#   implementation has some concerning details (e.g. busy waiting) for high scale
+#   usage. Given the complexity of the code base, we decided not to try to fork it
+#   to add the needed features.
+
+require 'aws-sdk'
+require 'logger'
+require 'thread'
+require_relative 'messages_handler'
+require_relative 'metrics'
+require_relative 'queue_processor_config'
+require_relative 'rate_limiter'
+
+module SQS
+
+  # A class for processing an SQS queue using a pool of worker threads, each of which
+  # does long polling against the queue.
+  class QueueProcessor
+    attr_reader :state, :logger, :config, :metrics, :handler
+
+    # @param [SQS::QueueProcessorConfig] config
+    # @param [SQS::Metrics] metrics
+    def initialize(config, metrics)
+      raise ArgumentError if config.nil? || metrics.nil?
+      @state = :initialized
+      @config = config
+      @metrics = metrics
+      @logger = config.logger
+      @handler = config.handler
+    end
+
+    # Begin processing messages. Must be called exactly once after initialize and before
+    # stop is called. This does not block and returns immediately after starting the worker
+    # thread, so be sure to make to keep the process running.
+    def start
+      assert_state :initialized, "Can't run in state #{@state}, must be :initialized"
+      @state = :running
+
+      # Kill the process if a thread dies with an unhandled exception. We attempt to catch all possible
+      # exceptions, so this shouldn't happen.
+      Thread.abort_on_exception = true
+      @worker_threads = []
+
+      (1..@config.num_workers_per_processor).each do |i|
+        worker_thread = Thread.new do
+          Thread.current[:name] = "SQS worker #{i}"
+          rate_limiter = RateLimiter.new(@config)
+          poller = Aws::SQS::QueuePoller.new(@config.queue_url)
+
+          # Break out of the polling loop when we leave the running state.
+          poller.before_request do |stats|
+            throw :stop_polling if @state != :running
+          end
+
+          # Wrap the polling loop in an outer exception handler and while loop,
+          # to prevent an unanticipated exception from terminating the thread.
+          while @state == :running
+            begin
+              # Long-poll for messages and handle them until we're told to stop.
+              poller.poll(max_number_of_messages: 10, wait_time_seconds: 20, visibility_timeout: 5) do |sqs_messages|
+                batch_failed = false
+                messages = sqs_messages.map {|sqs_message|
+                  SQS::Message.new(sqs_message.body)
+                }
+                batch_size = sqs_messages.size
+
+                start_time_sec = Time.now.to_f
+                begin
+                  @handler.handle(messages)
```

Comment:
ah of course, those are very good reasons to pass the full batch array through to the handler.